resultstoreui.py

```diff
@@ -0,0 +1,510 @@
+import grpc
+import contextlib
+import uuid
+import logging
+import os
+import time
+from absl import (flags, app)
+from google import auth
+from google.auth.transport import grpc as google_auth_transport_grpc
+from google.auth.transport import requests as google_auth_transport_requests
+
+from resultstoreapi.cloud.devtools.resultstore_v2.proto import (
+    resultstore_upload_pb2_grpc, resultstore_upload_pb2,
+    resultstore_download_pb2_grpc, resultstore_download_pb2, invocation_pb2,
+    invocation_pb2_grpc, action_pb2, common_pb2, configured_target_pb2,
+    target_pb2)
+from resultstoreui_utils import (get_default_invocation,
+                                 get_default_configuration, gen_new_uuid,
+                                 get_parent, get_default_target,
+                                 get_default_configured_target,
+                                 get_default_action, upload_files_to_bigstore,
+                                 get_name)
+from google.protobuf import (duration_pb2)
+
+_SERVER_ADDR_TEMPLATE = 'resultstore.googleapis.com'
+_FIELD_MASK_HEADER = 'x-goog-fieldmask'
+
+_LOGGER = logging.getLogger(__name__)
+_LOGGER.setLevel(logging.INFO)
+
+FLAGS = flags.FLAGS
+flags.DEFINE_string('command', '', 'Command to execute')
+flags.DEFINE_string('config_id', 'default', 'Config Id')
+flags.DEFINE_string('target_name', '', 'Target Name')
+flags.DEFINE_string('invocation_id', '', 'Invocation ID')
+flags.DEFINE_string('authorization_token', '', 'Authorization Token')
+flags.DEFINE_string('action_type', 'TEST', 'Action Type')
+flags.DEFINE_string('invocation_name', '', 'Invocation Name')
+flags.DEFINE_enum('status', 'PASSED', common_pb2.Status.keys(),
+                  'Status of this action (if command not provided)')
+flags.DEFINE_string('bucket_name', 'sival-logs',
+                    'Google Cloud Storage Bucket Name')
+flags.DEFINE_list('files', [], 'files to be uploaded with the action')
+
+
+class Error(Exception):
+    """Generic Exception class for this module."""
+
+
+def get_authorization_token():
+    """
+    Generates an authoirzation token if the user does not specify one via
+    the --authorization_token flag
+
+    Returns:
+        An authorization token
+    """
+    if not FLAGS.authorization_token:
+        authorization_token = gen_new_uuid()
+        _LOGGER.error('Your auth token is: %s', authorization_token)
+        return authorization_token
+    return FLAGS.authorization_token
+
+
+@contextlib.contextmanager
+def create_oauth_client(addr):
+    """
+    Creates a secure channel using GOOGLE_APPLICATION_CREDENTIALS from the
+    users path
+
+    Args:
+        target (str): The host and port of the service
+    
+    Returns:
+        A gRPC channel
+    """
+    credentials, _ = auth.default(
+        scopes=("https://www.googleapis.com/auth/cloud-source-tools",
+                "https://www.googleapis.com/auth/cloud-platform"))
```

Comment:
Make these scopes a global constant.

Combine with bigstore scopes below in a common credentials file.